### ISSUE TITLE
Closing run loops in multi threaded example

### ISSUE THREAD

**Issue by Nevergonnachange** [uWebSockets/uWebSockets#366]
Hello, I have a short question:

What operating system are you running: Windows
What platform/language are you using: C++
Are you using SSL, if so, how: SSL is not part of the equation

**How can I close the run loop of the multi thread example. (A code example for this would be awesome)**

`threadedServer[i]->getDefaultGroup<uWS::SERVER>().addAsync();`
`threadedServer[i]->run();`

Thank you very much in advance

Regards
Robin

**Comment by alexhultman** [comment]
If you do threading you have to look at threading with libuv. Hub.getLoop gives you the uv_loop_t which can be used to fire uv_async_t which is like a thread safe delegation. You have a close function of Group which you can call inside the uv_async_t callback. Also consider if you really need threading when you have libuv.

**Comment by kolis** [comment]
For reference, this is how I solved it
```cpp
  auto uv_async_callback = [](uv_async_t* handle) {
    auto uv_walk_callback = [](uv_handle_t* handle, void* /*arg*/) {
      if (!uv_is_closing(handle))
          uv_close(handle, nullptr);
    };

    auto loop = handle->loop;
    uv_stop(loop);
    uv_walk(loop, uv_walk_callback, nullptr);
    uv_run(loop, UV_RUN_DEFAULT);
    uv_loop_close(loop);
  };

  auto loop = m_hub.getLoop();

  uv_async_t asyncHandle;
  uv_async_init(loop, &asyncHandle, uv_async_callback);
  uv_async_send(&asyncHandle);
```

I was helped a lot by this post http://stackoverflow.com/a/25831688/1966638

**Comment by alexhultman** [comment]
This is not correct. You have Group::close() which closes connections and stops listening and makes the loop fall through. Hub::getDefaultGroup<SERVER>().close();

**Comment by alexhultman** [comment]
Out of curiosity though - why would anyone want to end a server? To me it sounds strange given that a server should run forever.

**Comment by kolis** [comment]
Ok, thanks for feedback!

Why we want to end the server, well, I'm running a RPC interface through uWS in a desktop application.

**Comment by alexhultman** [comment]
Ah! Makes sense. I would probably just exit(0); if I got frustrated hehe

**Comment by ad34** [comment]
@kolis thanks for posting your code. on my side I use hub.run() in a separate thread. (uWebSockets as client in C++ app) I manage to fix a 0.6 MB memory leaks adding a async closing on app exit.
